### PR TITLE
Outlier reports

### PR DESCRIPTION
**Pull request by Jongmassey**
Migration of https://github.com/ebmdatalab/outliers/pull/42 to openprescribing management command

### CODE REVIEW

File: openprescribing/pipeline/management/commands/outlier_reports.py [hunk 1]
```diff
@@ -0,0 +1,1656 @@
...
+        self.entity_hierarchy: Dict[str, Dict[str, List[str]]] = {}
+
+    def run(self) -> None:
+        """
+        Execute outlier dataset build stored procedure
+
+        Populates build_id attribute upon completion
+        """
+        sql = f"""
+            CALL `ebmdatalab.outlier_detection.build_outliers`(
```
Comment (inglesp) [inline openprescribing/pipeline/management/commands/outlier_reports.py:529]: Is the code for this stored procedure checked in anywhere?

File: openprescribing/pipeline/management/commands/outlier_reports.py [hunk 2]
```diff
@@ -0,0 +1,1656 @@
...
+        self.build.run()
+        self.build.fetch_results()
+        self._truncate_entities()
+        self._truncate_results()
+        self.toc.hierarchy = self.build.entity_hierarchy
+        self.run_results = {e: [] for e in self.build.entities}
+
+        # loop through entity types, generated a report for each entity item
+        for e in self.build.entities:
+            for f in self._run_entity_report(e):
```
Comment (inglesp) [inline openprescribing/pipeline/management/commands/outlier_reports.py:1208]: What's `f` here?

File: openprescribing/pipeline/management/commands/outlier_reports.py (hunk 1 above)
Comment (Jongmassey) [reply openprescribing/pipeline/management/commands/outlier_reports.py:537]: https://github.com/ebmdatalab/outliers/tree/master/data/static_outlier_sql

File: openprescribing/pipeline/management/commands/outlier_reports.py [hunk 3]
```diff
@@ -0,0 +1,1657 @@
...
+                    else:
+                        continue
+                    break
+            one_practice_per_pcn()
+            return
+
+        if self.entity_limit < practice_count():
+            while True:
+                for _, ccgs in self.build.entity_hierarchy.items():
+                    for _, pcns in ccgs.values():
```
[the remark this answers is not in the document]
Comment (Jongmassey) [reply openprescribing/pipeline/management/commands/outlier_reports.py]: correct, the latter, I think I've already resolved this

File: openprescribing/pipeline/management/commands/outlier_reports.py [hunk 4]
```diff
@@ -0,0 +1,1657 @@
...
+                        v.values()
+                        for v in self.build.entity_hierarchy.values()
+                    ]
+                    for x in y
+                ]
+                for p in q
+            ]
+            for a in b
+        ]
+
```
[the remark this answers is not in the document]
Comment (Jongmassey) [reply openprescribing/pipeline/management/commands/outlier_reports.py:1397]: yeah, it's pretty ugly. I'll have a think about making it clearer.

File: openprescribing/pipeline/management/commands/outlier_reports.py [hunk 5]
```diff
@@ -0,0 +1,1657 @@
...
+    def __init__(
+        self,
+        from_date: date,
+        to_date: date,
+        n_outliers: int,
+        entities: List[str],
+        force_rebuild: bool = False,
+        entity_limit: int = None,
+        output_dir="../data",
+        template_path="../data/template.html",
```
[the remark this answers is not in the document]
Comment (Jongmassey) [reply openprescribing/pipeline/management/commands/outlier_reports.py:1184]: I think you're looking at an outdated version of the code, these have already been replaced.

File: openprescribing/pipeline/management/commands/outlier_reports.py [hunk 6]
```diff
@@ -0,0 +1,1657 @@
...
+            "entity": entity,
+            "file_path": output_file,
+        }
+
+    def _run_entity_report(self, entity):
+        codes = self.build.results[entity].index.get_level_values(0).unique()
+        kwargs = [{"entity": entity, "code": c} for c in codes]
+        with ThreadPoolExecutor(max_workers=self.n_jobs) as pool:
+            futures = [
+                pool.submit(self._run_entity_report, **k) for k in kwargs
```
[the remark this answers is not in the document]
Comment (Jongmassey) [reply openprescribing/pipeline/management/commands/outlier_reports.py]: >should it be pool.submit(self._run_item_report, ...)?

very yes! Again this has already been changed

**Pull request merged**